lib/opal/rewriters/logical_operator_assignment.rb

```diff
@@ -0,0 +1,199 @@
+          lhs.updated(:cvasgn, [*lhs, rhs]) # @@lhs = rhs
+        end
+      end
+
+      # Takes    `recvr.meth ||= rhs`
+      # Produces `recvr.meth || recvr.meth = rhs`
+      # (lhs is a recvr.meth)
+      class SendHandler < AbstractHandler
+        def self.process(lhs, rhs, root_type)
+          recvr, reader_method, *args = *lhs
+
+          # If recvr is a complex expression it must be cached.
+          # MRI calls recvr in `recvr.meth ||= rhs` only once.
+          if recvr && recvr.type == :send
+            recvr_tmp = new_temp
+            cache_recvr = s(:lvasgn, recvr_tmp, recvr)                         # $tmp = recvr
+            recvr = s(:js_tmp, recvr_tmp)
+          end
+
+          writer_method = :"#{reader_method}="
+
+          call_reader = lhs.updated(:send, [recvr, reader_method, *args])      # $tmp.meth
+          call_writer = lhs.updated(:send, [recvr, writer_method, *args, rhs]) # $tmp.meth = rhs
+          get_or_set = s(root_type, call_reader, call_writer)
+
+          if cache_recvr
+            s(:begin, cache_recvr, get_or_set)
+          else
+            get_or_set
+          end
+        end
+      end
+
+      # Takes    `recvr&.meth ||= rhs`
+      # Produces `recvr.nil? ? nil : recvr.meth ||= rhs`
+      #   NOTE: Later output of this handler gets post-processed by this rewriter again
+      #   using SendHandler to `recvr.nil? ? nil : (recvr.meth || recvr.meth = rhs)`
+      class ConditionalSendHandler < AbstractHandler
+        def self.process(lhs, rhs, root_type)
+          root_type = :"#{root_type}_asgn"
+
+          recvr, meth, *args = *lhs
+
+          recvr_tmp = new_temp
+          cache_recvr = s(:lvasgn, recvr_tmp, recvr)          # $tmp = recvr
+          recvr = s(:js_tmp, recvr_tmp)
+
+          recvr_is_nil = s(:send, recvr, :nil?)                 # recvr.nil?
+          nil_node = s(:nil)                                    # nil
+          plain_send = lhs.updated(:send, [recvr, meth, *args]) # recvr.meth
+          plain_or_asgn = s(root_type, plain_send, rhs)         # recvr.meth ||= rhs
+
+          s(:begin,
+            cache_recvr,
+            s(:if, recvr_is_nil,                          # if recvr.nil?
+              s(:nil),                                    #   nil
+                                                          # else
+              plain_or_asgn))                             #   recvr.meth ||= rhs
+                                                          # end
+        end
+      end
+
+      HANDLERS = {
+        lvasgn: LocalVariableHandler,
+        ivasgn: InstanceVariableHandler,
+        casgn:  ConstantHandler,
+        gvasgn: GlobalVariableHandler,
+        cvasgn: ClassVariableHandler,
+        send:   SendHandler,
+        csend:  ConditionalSendHandler
+      }
+
+      # lhs ||= rhs
+      def on_or_asgn(node)
+        lhs, rhs = *node
+
+        result = HANDLERS
+          .fetch(lhs.type) { raise NotImplementedError }
```

Comment:
Yeah, I would better raise a `Parser::SyntaxError` here, but there's no real loc that could be pointed by this exception. The error may come *only* from other broken rewriter that could be placed before in the chain of `Opal::Rewriter.list`. If the grammar is broken `parser` shows it. But generally - yes, we have enough custom exceptions in `opal/nodes`. Probably we could use `parser.diagnostic` to properly display those errors.